feat: playbooks

adityathebe opened the pull request.

moshloop commented on playbook/controllers.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// HandlePlaybookRun handles playbook run requests.
+func HandlePlaybookRun(c echo.Context) error {
```
I think this should queue the playbook run, with a startdate of now and then we have a background process that runs them

moshloop commented on playbook/controllers.go:
```diff
@@ -0,0 +1,134 @@
+		return c.JSON(http.StatusBadRequest, api.HTTPError{Error: err.Error(), Message: "invalid request"})
+	}
+
+	playbook, err := db.FindPlaybook(ctx, req.ID)
```
I think the execution of the playbook should be in a separate file/function - especially so that it can be tested.

adityathebe replied:
The function that executes is isolated. This controller simply records the run in the db.

adityathebe commented:
@moshloop This is ready for review now.
- implemented approvals by a person/team
- endpoint that lists all the suitable playbooks for a config/component

adityathebe commented:
@moshloop @yashmehrotra I've added a new application level error struct that can be easily mapped to HTTP errors. It was kind of necessary for me right now so I've added it in this PR itself. 

It's pretty much a copy of https://github.com/benbjohnson/wtf/blob/main/error.go with some minor changes.

moshloop commented on playbook/playbook.go:
```diff
@@ -0,0 +1,30 @@
+		return nil, api.Errorf(api.ENOTFOUND, "component(id=%s) not found", id)
+	}
+
+	return db.FindPlaybooksByTypeAndTags(ctx, component.Type, component.Labels)
```
I think this will. need to be another function and look at components and not configs

moshloop commented on playbook/playbook_test.go:
```diff
@@ -0,0 +1,201 @@
+		}
+	})
+
+	ginkgo.It("should execute playbook", func() {
```
Can we add tests for playbook selection on both components and configs

moshloop commented on playbook/playbook_test.go:
```diff
@@ -0,0 +1,201 @@
+	})
+
+	ginkgo.It("should store playbook run via API", func() {
+		run := RunParams{
```
I don't think it is necessary to call these functions via HTTP - we can all the functions directly

adityathebe replied:
Everything about saving a playbook run is handled by the HTTP handler. The actual task is simply to save to DB and the handler performs other validation on top of that.

I think its better to test the handler this way than just directly saving to db from the test.

moshloop commented on playbook/queue_consumer.go:
```diff
@@ -0,0 +1,159 @@
+
+func (t *queueConsumer) Listen() error {
+	pgNotify := make(chan string)
+	go utils.ListenToPostgresNotify(t.pool, "playbook_run_updates", t.dbReconnectMaxDuration, t.dbReconnectBackoffBaseDuration, pgNotify)
```
I would move this method to the QueueConsumer .e.g

t.PgListen("playbook_spec_approval", approvalUpdated)

moshloop commented on playbook/queue_consumer.go:
```diff
@@ -0,0 +1,159 @@
+	}
+}
+
+func (t *queueConsumer) onPlaybookSpecApprovalUpdated(ctx *api.Context, playbookID string) error {
```
These functions are all in the playbook package so no need to duplicate prefix
```suggestion
func (t *queueConsumer) onApprovalUpdated(ctx *api.Context, playbookID string) error {
```

moshloop commented on playbook/queue_consumer.go:
```diff
@@ -0,0 +1,96 @@
+	for _, r := range runs {
+		go func(run models.PlaybookRun) {
+			if _, loaded := t.registry.LoadOrStore(run.ID, nil); !loaded {
+				if !run.StartTime.After(time.Now()) {
```
I don't think we should be sleeping inside go functions - we also shouldn't be comparing server time and db time. 

What is wrong with `select * from runs where starttime >= now() and status = 'pending'` ?

Can we not re-use the queue consumer for events ? as it is a very similar process, just on a different table

adityathebe replied:
> select * from runs where starttime >= now() and status = 'pending'

In this case we could be fetching runs which are schedule in future. And it wouldn't execute runs whose start time was missed.

adityathebe replied:
The scheduled time doesn't have to be accurate. It's fine to have an offset of a minute.

moshloop commented on api/errors.go:
```diff
@@ -0,0 +1,86 @@
+
+// ErrorCode unwraps an application error and returns its code.
+// Non-application errors always return EINTERNAL.
+func ErrorCode(err error) string {
```
This should probably be a WithError

adityathebe replied:
`WithError` would make sense if we were attaching the error. But here we're just extracting the code from the error.

moshloop commented on playbook/runner.go:
```diff
@@ -0,0 +1,137 @@
+
+func executeAction(ctx *api.Context, run models.PlaybookRun, action v1.PlaybookAction, actionParam ActionParam) ([]byte, error) {
+	if action.Exec != nil {
+		script, err := gomplate.RunTemplate(actionParam.AsMap(), gomplate.Template{Template: action.Exec.Script})
```
I think this should be in the exec.go file

moshloop commented on playbook/runner.go:
```diff
@@ -0,0 +1,137 @@
+			return nil, err
+		}
+
+		return json.Marshal(res.Stdout)
```
I think we need returncode, stdout and stderr in a json object - and they need to be saved backed to the playbook_run_action.result column

moshloop commented on fixtures/playbooks/deleting-configmap.yaml:
```diff
@@ -0,0 +1,15 @@
+  configs:
+    - type: Kubernetes::ConfigMap
+      tags:
+        namespace: default
```
namespace should come from the config itself
